### PR TITLE
Convert MethodParameterResolver to generic ParameterResolver API

### PR DESCRIPTION
**Pull request by sbrannen**
Prior to this commit, the `MethodParameterResolver` API could only be applied to methods. In addition, the `MethodInvocationContext` API needlessly stored the method which could be retrieved directly from the supplied `Parameter`.

This commit addresses these issues by converting the `MethodParameterResolver` into a generic `ParameterResolver` API that can be used to resolve parameters for both methods and constructors. At the moment, however, there is no first-class support for resolving constructor parameters in the `JUnit5TestEngine`.

Specifically, this commit includes the following changes.

- Renamed `MethodParameterResolver` to `ParameterResolver`
- Replaced `MethodInvocationContext` arguments in `ParameterResolver` with `Optional<Object>` for the target of method invocations
- Moved `MethodInvocationContext` to the `junit5-engine` module for internal use

Issue: #267

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

File: junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+@API(Experimental)
+public interface ParameterResolver extends Extension {
+
+	/**
+	 * Determine if this resolver supports resolution of the given {@link Parameter}
+	 * for the supplied {@code target} and {@link ExtensionContext}.
+	 *
+	 * <p>The {@link java.lang.reflect.Method} or {@link java.lang.reflect.Constructor}
+	 * in which the {@code parameter} is declared can be retrieved via
+	 * {@link Parameter#getDeclaringExecutable()}.
```
Comment (marcphilipp) [inline junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java]: 👍

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java [hunk 2]
```diff
@@ -73,18 +73,21 @@ public Object invoke(MethodInvocationContext methodInvocationContext) {
 	private Object resolveParameter(Parameter parameter, MethodInvocationContext methodInvocationContext)
 			throws ParameterResolutionException {
 
-		Method method = methodInvocationContext.getMethod();
+		final Method method = methodInvocationContext.getMethod();
+		final Optional<Object> target = Optional.ofNullable(methodInvocationContext.getInstance());
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: We usually don't declare local variables `final`, do we?

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java [hunk 3]
```diff
@@ -73,18 +73,21 @@ public Object invoke(MethodInvocationContext methodInvocationContext) {
...
 
 		try {
 			// @formatter:off
-			List<MethodParameterResolver> matchingResolvers = this.extensionRegistry.stream(MethodParameterResolver.class)
-					.filter(resolver -> resolver.supports(parameter, methodInvocationContext, this.extensionContext))
+			List<ParameterResolver> matchingResolvers = this.extensionRegistry.stream(ParameterResolver.class)
+					.filter(resolver -> {
+						return resolver.supports(parameter, target, this.extensionContext);
+					})
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: Did you change this into a block for better readability?

**Comment by marcphilipp** [comment]
I think we should add a test case to `ParameterResolverTests` that checks that `ParameterResolver` can be used with a test class constructor.

File: junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java (hunk 1 above)
Comment (sbrannen) [inline junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java]: I take it that means you were not aware of that method. 😉

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 2 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: no, not usually. I'll change that.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 3 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: Ummmm.... I don't recall having done that.

Strange. I'll change it back.

**Comment by sbrannen** [comment]
> I think we should add a test case to `ParameterResolverTests` that checks that `ParameterResolver` can be used with a test class constructor.

I'll see what I can cook up.

File: junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java (hunk 1 above)
Comment (marcphilipp) [inline junit5-api/src/main/java/org/junit/gen5/api/extension/ParameterResolver.java]: Not until you told me a few days ago. 😉

I just meant it's good that you documented it here. 🙂

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 2 above)
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: Thanks!

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 3 above)
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: Thanks!

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 2 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: done in forthcoming forced push.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java (hunk 3 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/execution/MethodInvoker.java]: done in forthcoming forced push.

**Comment by marcphilipp** [comment]
Looks good! Feel free to merge when you've added the additional test.

**Pull request merged**